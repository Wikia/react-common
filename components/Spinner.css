@keyframes fandom-spinner-rotator {
  from {
    transform: rotate(0);
  }
  to {
    transform: rotate(360deg);
  }
}
@keyframes fandom-spinner-dash {
  to {
    stroke-dashoffset: 0;
  }
}
.fandom-spinner {
  display: inline-flex;
}
.fandom-spinner > svg {
  animation: fandom-spinner-rotator 0.5s linear infinite;
  transform: translateZ(0);
}
.fandom-spinner > svg circle {
  animation: fandom-spinner-dash 1.25s linear infinite alternate-reverse;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  stroke: #088488;
}
.fandom-spinner.is-block {
  display: block;
  left: auto;
  margin: auto;
  position: relative;
  text-align: center;
}
.fandom-spinner.is-inline {
  height: 1em;
  left: 0;
  margin: 0;
  position: static;
  top: 0;
  width: auto;
}