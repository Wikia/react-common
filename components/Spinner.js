'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var PropTypes = _interopDefault(require('prop-types'));

/**
 * Loader block component used to indicate loading state.
 *
 * Based on http://fandomdesignsystem.com/#/components/progress-indicators
 */

var Spinner = function Spinner(_ref) {
  var className = _ref.className,
      size = _ref.size,
      stroke = _ref.stroke,
      block = _ref.block,
      inline = _ref.inline;
  var style = {
    width: size,
    height: size
  };
  var r = (size - stroke) / 2;
  var translate = r + stroke / 2;
  var dash = 2 * Math.PI * r;
  var classes = ['fandom-spinner'];

  if (block) {
    classes.push('is-block');
  }

  if (inline) {
    classes.push('is-inline');
  }

  if (className) {
    classes.push(className);
  }

  return React.createElement("div", {
    className: classes.join(' '),
    style: style
  }, React.createElement("svg", {
    width: size,
    height: size,
    viewBox: "0 0 ".concat(size, " ").concat(size),
    xmlns: "http://www.w3.org/2000/svg"
  }, React.createElement("g", {
    transform: "translate(".concat(translate, ", ").concat(translate, ")")
  }, React.createElement("circle", {
    fill: "none",
    strokeWidth: stroke,
    strokeDasharray: dash,
    strokeDashoffset: dash,
    strokeLinecap: "round",
    r: r
  }))));
};

Spinner.propTypes = {
  /**
  * Additional class name
  */
  block: PropTypes.bool,

  /**
  * Loader size
  */
  className: PropTypes.string,

  /**
  * Stroke width
  */
  inline: PropTypes.bool,

  /**
  * Display block and center
  */
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),

  /**
  * Display contentinline based on line height
  */
  stroke: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};
Spinner.defaultProps = {
  className: '',
  size: 30,
  stroke: 2,
  block: false,
  inline: false
};

module.exports = Spinner;
