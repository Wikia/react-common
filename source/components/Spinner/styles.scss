// import variables
@import "~design-system/dist/scss/wds-variables/index.scss";

@keyframes fandom-spinner-rotator {
	from {
		transform: rotate(0);
	}

	to {
		transform: rotate(360deg);
	}
}
@keyframes fandom-spinner-dash {
	to {
		stroke-dashoffset: 0;
	}
}

.fandom-spinner {
	display: inline-flex;

	& > svg {
		animation: fandom-spinner-rotator .5s linear infinite;
		transform: translateZ(0);

		circle {
			animation: fandom-spinner-dash 1.25s linear infinite alternate-reverse;
			-webkit-backface-visibility: hidden;
			backface-visibility: hidden;
			stroke: $wds-color-link;
		}
	}

	&.is-block {
		display: block;
		left: auto;
		margin: auto;
		position: relative;
		text-align: center;
	}

	&.is-inline {
		height: 1em;
		left: 0;
		margin: 0;
		position: static;
		top: 0;
		width: auto;
	}
}
