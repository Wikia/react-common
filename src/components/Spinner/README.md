Defaults:
```js
<Spinner />
```

Custom size and stroke:
```js
<div>
	<Spinner size={'78'} stroke={'6'}/>
	<Spinner size={70} stroke={2}/>
</div>
```
