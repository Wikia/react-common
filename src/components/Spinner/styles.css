@keyframes fandom-spinner-rotator {
	from {
		transform: rotate(0);
	}

	to {
		transform: rotate(360deg);
	}
}
@keyframes fandom-spinner-dash {
	to {
		stroke-dashoffset: 0;
	}
}

.fandom-spinner {
	display: inline-flex;
}

.fandom-spinner .wds-spinner {
	animation: fandom-spinner-rotator .5s linear infinite;
	transform: translateZ(0);
}

.fandom-spinner .wds-spinner__stroke {
	animation: fandom-spinner-dash 1.25s linear infinite alternate-reverse;
	-webkit-backface-visibility: hidden;
	backface-visibility: hidden;
}
