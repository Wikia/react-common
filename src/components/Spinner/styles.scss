// import variables
@import "~design-system/dist/scss/wds-variables/index.scss";

@keyframes fandom-spinner-rotator {
	from {
		transform: rotate(0);
	}

	to {
		transform: rotate(360deg);
	}
}
@keyframes fandom-spinner-dash {
	to {
		stroke-dashoffset: 0;
	}
}

.fandom-spinner {
	display: inline-flex;

	& > svg {
		animation: fandom-spinner-rotator .5s linear infinite;
		transform: translateZ(0);

		circle {
			animation: fandom-spinner-dash 1.25s linear infinite alternate-reverse;
			-webkit-backface-visibility: hidden;
			backface-visibility: hidden;
			stroke: $wds-color-link;
		}
	}
}
