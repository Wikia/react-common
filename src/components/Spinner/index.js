import React from 'react';
import PropTypes from 'prop-types';

import colors from '../../../utils/colors';

import './styles.scss';

/**
 * Loader block component used to indicate loading state.
 *
 * Based on http://fandomdesignsystem.com/#/components/progress-indicators
 */
const Spinner = ({
  className,
  color,
  size,
  stroke,
}) => {
  const style = {
    width: size,
    height: size,
  };

  const r = (size - stroke) / 2;
  const translate = r + (stroke / 2);
  const dash = 2 * Math.PI * r;

  return (
    <div className={`fandom-spinner ${className}`} style={style}>
      <svg
        width={size}
        height={size}
        viewBox={`0 0 ${size} ${size}`}
        xmlns="http://www.w3.org/2000/svg"
      >
        <g transform={`translate(${translate}, ${translate})`}>
          <circle
            fill="none"
            strokeWidth={stroke}
            strokeDasharray={dash}
            strokeDashoffset={dash}
            strokeLinecap="round"
            stroke={color}
            r={r}
          />
        </g>
      </svg>
    </div>
  );
};

Spinner.propTypes = {
  /**
   * Additional class name
   */
  className: PropTypes.string,
  /**
   * Loader size
   */
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Stroke width
   */
  stroke: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * Stroke color
   */
  color: PropTypes.string,
};

Spinner.defaultProps = {
  className: '',
  size: 30,
  stroke: 2,
  color: colors.WDS_COLOR_LINK,
};

export default Spinner;
