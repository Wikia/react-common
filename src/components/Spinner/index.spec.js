import React from 'react';
import renderer from 'react-test-renderer';

import Spinner from './index';

test('Spinner renders correctly with default values', () => {
    const component = renderer.create(
        <Spinner />,
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Spinner renders correctly with custom size (string)', () => {
    const component = renderer.create(
        <Spinner size="78" stroke="6" />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Spinner renders correctly with custom size (number)', () => {
    const component = renderer.create(
        <Spinner size="20" />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Spinner renders correctly with custom colors (named color)', () => {
    const component = renderer.create(
        <Spinner color="red" />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Spinner renders correctly with custom colors (short hex)', () => {
    const component = renderer.create(
        <Spinner color="#0f0" />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Spinner renders correctly with custom colors (rgb)', () => {
    const component = renderer.create(
        <Spinner color="rgb(0,0,255)" />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Spinner renders correctly with custom inline ', () => {
    const component = renderer.create(
        <Spinner inline />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Spinner renders correctly with custom block ', () => {
    const component = renderer.create(
        <Spinner block />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test('Spinner renders correctly with additional classname', () => {
    const component = renderer.create(
        <Spinner className="custom-class or-multiple-classes" />
    );
    expect(component.toJSON()).toMatchSnapshot();
});
