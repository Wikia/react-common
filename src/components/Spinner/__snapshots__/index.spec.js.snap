// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner renders correctly with custom colors (named color) 1`] = `
<div
  className="fandom-spinner "
  style={
    Object {
      "height": 30,
      "width": 30,
    }
  }
>
  <svg
    height={30}
    viewBox="0 0 30 30"
    width={30}
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(15, 15)"
    >
      <circle
        fill="none"
        r={14}
        strokeDasharray={87.96459430051421}
        strokeDashoffset={87.96459430051421}
        strokeLinecap="round"
        strokeWidth={2}
      />
    </g>
  </svg>
</div>
`;

exports[`Spinner renders correctly with custom colors (rgb) 1`] = `
<div
  className="fandom-spinner "
  style={
    Object {
      "height": 30,
      "width": 30,
    }
  }
>
  <svg
    height={30}
    viewBox="0 0 30 30"
    width={30}
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(15, 15)"
    >
      <circle
        fill="none"
        r={14}
        strokeDasharray={87.96459430051421}
        strokeDashoffset={87.96459430051421}
        strokeLinecap="round"
        strokeWidth={2}
      />
    </g>
  </svg>
</div>
`;

exports[`Spinner renders correctly with custom colors (short hex) 1`] = `
<div
  className="fandom-spinner "
  style={
    Object {
      "height": 30,
      "width": 30,
    }
  }
>
  <svg
    height={30}
    viewBox="0 0 30 30"
    width={30}
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(15, 15)"
    >
      <circle
        fill="none"
        r={14}
        strokeDasharray={87.96459430051421}
        strokeDashoffset={87.96459430051421}
        strokeLinecap="round"
        strokeWidth={2}
      />
    </g>
  </svg>
</div>
`;

exports[`Spinner renders correctly with custom size (number) 1`] = `
<div
  className="fandom-spinner "
  style={
    Object {
      "height": "20",
      "width": "20",
    }
  }
>
  <svg
    height="20"
    viewBox="0 0 20 20"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(10, 10)"
    >
      <circle
        fill="none"
        r={9}
        strokeDasharray={56.548667764616276}
        strokeDashoffset={56.548667764616276}
        strokeLinecap="round"
        strokeWidth={2}
      />
    </g>
  </svg>
</div>
`;

exports[`Spinner renders correctly with custom size (string) 1`] = `
<div
  className="fandom-spinner "
  style={
    Object {
      "height": "78",
      "width": "78",
    }
  }
>
  <svg
    height="78"
    viewBox="0 0 78 78"
    width="78"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(39, 39)"
    >
      <circle
        fill="none"
        r={36}
        strokeDasharray={226.1946710584651}
        strokeDashoffset={226.1946710584651}
        strokeLinecap="round"
        strokeWidth="6"
      />
    </g>
  </svg>
</div>
`;

exports[`Spinner renders correctly with default values 1`] = `
<div
  className="fandom-spinner "
  style={
    Object {
      "height": 30,
      "width": 30,
    }
  }
>
  <svg
    height={30}
    viewBox="0 0 30 30"
    width={30}
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(15, 15)"
    >
      <circle
        fill="none"
        r={14}
        strokeDasharray={87.96459430051421}
        strokeDashoffset={87.96459430051421}
        strokeLinecap="round"
        strokeWidth={2}
      />
    </g>
  </svg>
</div>
`;
